cpp/dolfin/mesh/DistributedMeshTools.cpp

```diff
@@ -1172,3 +1181,15 @@ EigenRowArrayXXd DistributedMeshTools::reorder_values_by_global_indices(
   return new_values;
 }
 //-----------------------------------------------------------------------------
+template EigenRowArrayXXd
+DistributedMeshTools::reorder_values_by_global_indices<double>(
+    MPI_Comm mpi_comm,
+    const Eigen::Ref<const EigenRowArrayXXd>& values,
+    const std::vector<std::int64_t>& global_indices);
+//-----------------------------------------------------------------------------
+template Eigen::Array<std::complex<double>, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
+DistributedMeshTools::reorder_values_by_global_indices<std::complex<double>>(
+    MPI_Comm mpi_comm,
+    const Eigen::Ref<const Eigen::Array<std::complex<double>, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>& values,
+    const std::vector<std::int64_t>& global_indices);
+//-----------------------------------------------------------------------------
```

Comment:
This explicit instantiation in the `cpp` file looks strange to me and is confusing. The declaration is a template, but it looks like it's relying on some dodgy link-time checks to make things work.

As a public function, it should be generic (in which case the implementation needs to go into the header file), or have concrete types.

It's also a bit confusing see the typedef for `EigenRowArrayXXd`, and then a full declaration for the complex vase.

Reply:
@garth-wells Explicit instantiation is indeed not needed in this case, since compiler knows what functions to compile from their usage across .cpp files. However, as you say, when you try to link your own code with the method templated by type not used (e.g. eigen array of int) it'll fail the linking. 
There are at least 3 ways to handle this: 
1.) Expose template definition in header.
2.) Make some function a concrete implementation of `reorder_values_by_global_indices` and make `reorder_values_by_global_indices` private - as done in XDMFFile::write(meshfunction).
3.) Just template declaration in header, template definition in source and compiler deduction of type from its usage.

I like the most 3., since there are just 2 types to use with and method is not meant to be heavily used by users, but its your decision.

Ad `EigenRowArrayXXd`: I think @IgorBaratta suggested to typedef something like `EigenRowArrayXXc` for complex and make it look more symmetric, but you didn't want to add another typedef.